global_planner/src/planner_core.cpp

```diff
@@ -308,6 +308,47 @@ uint32_t GlobalPlanner::makePlan(const geometry_msgs::PoseStamped& start, const
     bool found_legal = planner_->calculatePotentials(costmap_->getCharMap(), start_x, start_y, goal_x, goal_y,
                                                     nx * ny * 2, potential_array_);
 
+
+    geometry_msgs::PoseStamped best_pose = goal;
+    if (!found_legal) {
+      double resolution = costmap_->getResolution();
+      geometry_msgs::PoseStamped p = goal;
+
+      double best_sdist = DBL_MAX;
+
+      unsigned int mx, my;
+      for(double dy = -tolerance; dy <= tolerance; dy += resolution){
+        p.pose.position.y = goal.pose.position.y + dy;
+        const double dx = std::sqrt(tolerance*tolerance - dy*dy);
+        for(p.pose.position.x = goal.pose.position.x - dx; p.pose.position.x <= goal.pose.position.x + dx; p.pose.position.x += resolution){
+            if(costmap_->worldToMap(p.pose.position.x, p.pose.position.y, mx, my)) {
+              unsigned int index = my * nx + mx;
+              double potential = potential_array_[index];
+              double sdist = sq_distance(p, goal);
+              if(potential < POT_HIGH && sdist < best_sdist){
+                best_sdist = sdist;
+                best_pose = p;
+                found_legal = true;
+              }
+            }
+        }
+      }
+
+      if (found_legal) {
+        if (!costmap_->worldToMap(best_pose.pose.position.x, best_pose.pose.position.y, goal_x_i, goal_y_i)) {
```

Comment:
Q: why we need this since it was already tested before `found_legal` is set to true

Reply:
true